drivers/flash/flash_util.c

```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
 	return -ENOSYS;
 #endif
 }
+
+static bool off_add_overflow(off_t offset, off_t size, off_t *result)
+{
+	if (sizeof(off_t) == sizeof(uint32_t)) {
+		uint32_t end;
+
+		/* account for signedness of off_t due to lack of s32_add_overflow() */
+		if (u32_add_overflow((uint32_t)offset, (uint32_t)size, &end) ||
+		    (end > (UINT32_MAX >> 1))) {
+			return true;
+		}
+	} else if (sizeof(off_t) == sizeof(uint64_t)) {
+		uint64_t end;
+
+		/* account for signedness of off_t due to lack of s64_add_overflow() */
+		if (u64_add_overflow((uint64_t)offset, (uint64_t)size, &end) ||
+		    (end > (UINT64_MAX >> 1))) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+int flash_regions_overlap(off_t a_start, off_t a_size, off_t b_start, off_t b_size)
+{
+	off_t a_end = 0;
+	off_t b_end = 0;
+
+	BUILD_ASSERT((sizeof(off_t) == sizeof(uint32_t)) || (sizeof(off_t) == sizeof(uint64_t)));
+
+	if ((a_start < 0) || (a_size < 0) || (b_start < 0) || (b_size < 0)) {
+		return -EINVAL;
+	}
```

Comment:
That is why we need RFC for unsigned only `off_t` for storage, at least at the device level. I do not know any device that allows negative offsets to address bytes. Ok, I know one: that is sometimes in SPI NOR device, but it is rather side effect of overlap when crossing device size.

Reply:
> That is why we need RFC for unsigned only `off_t` for storage, at least at the device level. 

At the device level yes, but certain API calls (e.g. `lseek()`) do require the ability to seek backward, so a signed `off_t` is still necessary.

Reply:
Actually you do not if you have other parameter to indicate direction, something that `lseek` does not do

Reply:
`lseek()` doesn't have a direction parameter. It accepts `SEEK_SET`,  `SEEK_CUR`, and `SEEK_END` (for beginning, current position, and end) and then the `off_t` is relative to those locations.

Reply:
> `lseek()` doesn't have a direction parameter. It accepts `SEEK_SET`, `SEEK_CUR`, and `SEEK_END` (for beginning, current position, and end) and then the `off_t` is relative to those locations.

Yes, I have noticed
![image](https://github.com/user-attachments/assets/64f20f24-eb53-4e8f-b9ab-bf56990b983b)